### PR TITLE
[edward] Round-2: Progressive cosine EMA (0.99→0.9999) + RFF baseline

### PR DESCRIPTION
**Pull request by morganmcg1**
# [edward] Round-2: Progressive Cosine EMA (0.99→0.9999) + RFF Baseline

## Hypothesis

Progressive EMA decay schedule: start at `ema_decay=0.99` (fast/responsive, good for early training when model weights are changing rapidly) and ramp cosinely to `ema_decay=0.9999` (slow/stable, good for late training when weights have converged and noise reduction matters most).

This lever was the **single largest non-architectural gain in the yi branch**: PR #13 (norman, run `wio9pqw2`) achieved a ~10.3% improvement from yi baseline ~17.64 → 15.82 `test_abupt`. Tay currently uses a fixed `--ema-decay 0.9995` and has never tested this schedule. This experiment ports the yi win to tay, composed on top of fern's RFF features (PR #33).

**Why this matters for tay specifically:** With our limited epoch budget the model transitions quickly from exploration to convergence. A fixed EMA can't optimally serve both phases — too slow early (underweights recent updates), too fast late (adds noise). The progressive schedule gives the best of both worlds automatically.

## Current Tay SOTA (PR #33, fern — merged)

| Metric | tay SOTA (PR #33) | yi best (PR #13 EMA) | AB-UPT |
|---|---:|---:|---:|
| `abupt` | 17.77 | 15.82 | — |
| `surface_pressure` | 11.20 | 9.99 | 3.82 |
| `wall_shear` | 18.68 | 16.60 | 7.29 |
| `volume_pressure` | 16.13 | 14.21 | 6.08 |
| `tau_x` | 16.20 | 14.27 | 5.35 |
| `tau_y` | 21.81 | 19.49 | 3.65 |
| `tau_z` | 23.54 | 21.12 | 3.63 |

**Target:** `abupt < 17.77`. Expected lift: −1.0 to −1.8 (landing near 16.0 or below, approaching yi parity at 15.82).

## Implementation (train.py only, ~30 LOC)

Add the following to `train.py`:

### 1. New CLI arguments

```python
parser.add_argument("--ema-schedule", type=str, default="fixed",
                    choices=["fixed", "cosine", "linear"],
                    help="EMA decay schedule. 'fixed' uses --ema-decay unchanged (current behavior).")
parser.add_argument("--ema-decay-start", type=float, default=0.99,
                    help="Starting EMA decay for scheduled modes (default: 0.99).")
parser.add_argument("--ema-decay-end", type=float, default=0.9999,
                    help="Ending EMA decay for scheduled modes (default: 0.9999).")
```

### 2. Schedule computation (call at each optimizer step)

```python
import math

def get_ema_decay(schedule, step, total_steps, decay_start, decay_end, fixed_decay):
    """Return the EMA decay for the current step."""
    if schedule == "fixed":
        return fixed_decay
    t = min(step, total_steps)
    if schedule == "cosine":
        return decay_start + 0.5 * (decay_end - decay_start) * (1 - math.cos(math.pi * t / total_steps))
    elif schedule == "linear":
        return decay_start + (decay_end - decay_start) * (t / total_steps)
    return fixed_decay
```

### 3. Wire into training loop

At each step, before calling your EMA updater:

```python
if args.ema_schedule != "fixed":
    current_ema_decay = get_ema_decay(
        args.ema_schedule, global_step, total_steps,
        args.ema_decay_start, args.ema_decay_end, args.ema_decay
    )
    # Update EMA model's decay parameter if your EMA class supports it, e.g.:
    ema_model.decay = current_ema_decay
```

**Note:** Check how yi's student (norman) implemented the EMA update in `yi` branch `train.py` — specifically how `ema_model.decay` is updated mid-training. That implementation (run `wio9pqw2`, PR #13) is the reference. If your EMA class does not support updating `decay` in-place, refactor to pass it as an argument to the update step.

## Run Command

Single arm — run the cosine schedule directly on top of the RFF baseline:

```bash
torchrun --standalone --nproc_per_node=8 train.py \
  --volume-loss-weight 2.0 --batch-size 4 --validation-every 1 \
  --lr 5e-5 --weight-decay 5e-4 \
  --train-surface-points 65536 --eval-surface-points 65536 \
  --train-volume-points 65536 --eval-volume-points 65536 \
  --model-layers 4 --model-hidden-dim 512 --model-heads 8 --model-slices 128 \
  --rff-num-features 32 --rff-sigma 1.0 \
  --ema-schedule cosine --ema-decay-start 0.99 --ema-decay-end 0.9999 \
  --gradient-log-every 100 --weight-log-every 100 \
  --no-log-gradient-histograms --no-compile-model \
  --wandb-group edward-cosine-ema
```

All other hyperparameters inherit from the tay SOTA (PR #33) configuration.

## Expected Results

| Metric | tay SOTA | Expected (cosine EMA) |
|---|---:|---:|
| `abupt` | 17.77 | ~16.0 (−10% from yi precedent) |
| `surface_pressure` | 11.20 | ~10.1 |
| `wall_shear` | 18.68 | ~16.8 |

If the cosine schedule doesn't beat baseline but comes within 0.5, try: (a) a steeper start (0.999 instead of 0.99) or (b) a shorter warmup window (ramp over first 50% of steps only). Report both in your results comment.

## Deliverables

1. Report `val_abupt` at best epoch and final epoch.
2. Report the full metric table at best checkpoint (all 6 sub-metrics).
3. Include the W&B run ID.
4. Note the epoch at which the EMA decay crossed 0.999 (the midpoint) — this tells us how the schedule felt in practice.

## Suggested follow-ups (post-results)

- If cosine wins: try `--ema-decay-start 0.995 --ema-decay-end 0.9999` (narrower ramp, starts closer to current fixed value).
- If linear schedule vs cosine is curious: run a second arm with `--ema-schedule linear` (same start/end) and compare convergence curves in W&B.

### CODE REVIEW

**Comment by morganmcg1** [comment]
STUDENT edward:

## Implementation note + small deviation from PR command

Implemented `get_ema_decay()` per the PR + 4 new CLI flags: `--ema-schedule {fixed,cosine,linear}`, `--ema-decay-start`, `--ema-decay-end`, and an extra `--ema-schedule-total-steps`. Logging `train/ema_decay` to W&B every step.

**Why the extra flag:** Tay's 270-min train timeout caps runs at ~9 of the 50 max epochs (per project memory `project_run_timing.md` and confirmed on PR #33 fern run `u43lik5d` which logged 23,343 global steps). With the naive denominator of `max_epochs × steps_per_epoch ≈ 136,000` the cosine arc would only traverse ~17% of its range — EMA decay would crawl from 0.99 to ~0.991 instead of 0.99 → 0.9999. Same partial-arc bug as documented for `--lr-cosine-t-max 0`.

**Deviation:** Adding `--ema-schedule-total-steps 23000` to the launch command so the cosine arc actually completes over the realised tay budget. All other params unchanged from the PR command.

Debug verified on a 1-GPU tiny run (W&B `usagb6v5`): schedule=cosine, start=0.99, total_steps=200; `train/ema_decay` traversed 0.990001 → 0.999900 as expected.

Launching the full DDP8 run shortly.

**Comment by morganmcg1** [comment]
ADVISOR: Closing PR #44 because ADVISOR: Closing — cosine EMA schedule severely miscalibrated for our training budget; result 18.33 is worse than every prior baseline.

## Results

| Metric | PR #44 edward | PR #40 SOTA | PR #39 new SOTA | Δ vs PR #40 |
|---|---:|---:|---:|---:|
| test_abupt | **18.33** | 17.25 | **15.43** | **+6.3% regression** |
| surface_pressure | 11.69 | 10.92 | 9.45 | +7.1% regression |
| wall_shear | 19.35 | 18.33 | 16.28 | +5.6% regression |
| volume_pressure | 16.12 | 14.71 | 13.83 | +9.6% regression |

**This is worse than the uncompiled RFF run (PR #33, 17.77) and the calibration baseline (PR #30, 19.81 was better only because it was the first result).** Every axis regressed vs the compile-fix baseline.

## Root cause: EMA schedule maps to max_epochs=50, not to actual epochs run

The cosine schedule linearly sweeps ema_decay from 0.99 → 0.9999 over `ema_total_epochs` (which defaults to `max_epochs=50`). With our 270-min hard timeout we complete ~9-12 epochs, i.e. 18-24% of the schedule. At epoch 9 of 50:

```
ema_decay(9) ≈ 0.99 + (0.9999 - 0.99) × (1 - cos(π × 9/50)) / 2
             ≈ 0.99 + 0.0099 × 0.0967 ≈ 0.9910
```

Our fixed EMA baseline uses `ema_decay=0.9995`. Your cosine EMA spent the *entire run* at **0.9910-0.9925**, substantially below the fixed baseline — and yi's winning progressive EMA (PR #13) required reaching 0.9999 to get its win.

**The cosine EMA schedule needs `--ema-total-epochs` set to the actual planned epoch count (e.g., 12 for compiled runs), not the theoretical max.** As implemented, it's a strict downgrade for any budget < 50% of max_epochs.

## Yi context

Yi's norman PR #13 (cosine EMA, 15.82 test) was the largest non-architectural win on yi. But yi ran with a longer effective budget and a different compile situation. The mechanism is real — we just need to run it with T_max calibrated to our actual epoch count.

## What to try next

Given PR #39 (Lion) is now SOTA at 15.43, cosine EMA is worth composing with Lion at correct calibration:
- Base: Lion lr=1.7e-5, wd=5e-3
- Add: cosine EMA 0.99 → 0.9999 over `T_max = 12` (the compiled epoch count)
- Hypothesis: Lion's optimizer updates + EMA smoothing over the actual training horizon compound orthogonally

Reassigning edward to **Lion + RFF (single delta)** — sigma=1.0, 32 features, on top of the Lion winner — which is a cleaner first composition test. Cosine EMA composition can follow if Lion+RFF wins..
